compiler_gym/bin/manual_env.py

```diff
@@ -16,142 +16,590 @@
+
+                accept = not done and (eager_reward is not None) and (eager_reward > 0)
+                if accept:
+                    # Append the history element
+                    hist = ActionHistoryElement(
+                        action,
+                        index,
+                        eager_observation,
+                        eager_reward,
+                        done,
+                        info
                     )
+                    self.stack.append(hist)
+                else:
+                    # Basically undo
+                    self.rerun_stack()
 
-                # Print the eager reward and the diff, if available.
-                if env.reward_space and eager_reward is not None:
-                    reward_diff = ""
-                    if last_eager_reward is not None and eager_reward is not None:
-                        reward_diff = (
-                            f" (change: {eager_reward - last_eager_reward:.6f})"
-                        )
-                    print(f"Reward: {eager_reward:.6f}{reward_diff}")
-                    last_eager_reward = eager_reward
-
-                print(
-                    f"Action {env.action_space.names[index]} in {t}.",
-                    " No effect." if info.get("action_had_no_effect") else "",
-                    flush=True,
-                )
+                print(f"Step: {i+1} Action: {action} Reward: {eager_reward:.6f} Accept: {accept}")
                 if done:
                     print("Episode ended by environment: ", info["error_details"])
-                    env.close()
-                    return
-                break
-            if c == "o":
-                print("observation", flush=True)
-                observation_name = user_input.read_list_value(
-                    "Observable values", observation_names
+        print(f"Hill climbed {num_steps} steps in {timer}")
+
+    def get_action_rewards(self):
+        """Get all the rewards for the possible actions at this point"""
+        items = []
+        for index, action in enumerate(self.env.action_space.names):
+            self.rerun_stack()
+            eager_observation, eager_reward, done, info = self.env.step(index)
+            hist = ActionHistoryElement(
+                action,
+                index,
+                eager_observation,
+                eager_reward,
+                done,
+                info
+            )
+            items.append(hist)
+            print(f"Action: {action} Reward: {eager_reward:.6f}")
+
+        self.rerun_stack()
+        items.sort(key = lambda h: h.eager_reward, reverse = True)
+        return items
+
+    def do_try_all_actions(self, args):
+        """Tries all actions from this position and reports the results in sorted order by reward"""
+        with Timer() as timer:
+            items = self.get_action_rewards()
+        print(f"Got actions in {timer}")
```

Comment:
FYI (no action needed) there is a shortcut for this:

```suggestion
        with Timer("Got actions") as timer:
            items = self.get_action_rewards()
```

Reply:
Good one. I've changed those that can be changed.